From 3afae7e16930b07580b70ef9c36aad6219ab2dac Mon Sep 17 00:00:00 2001
From: Brendan King <Brendan.King@imgtec.com>
Date: Mon, 18 Jun 2018 16:15:55 +0100
Subject: [PATCH 06/16] glamor: add glsl shaders for OpenGLES 3.0
Upstream-Status: Pending

For those paths that require GLSL 130 with deskop OpenGL, add shaders
for OpenGLES 3.0 (glsl version 300), so that software fallback can be
avoided on OpenGLES 3.0.
---
 glamor/glamor.c                  |   3 +-
 glamor/glamor_composite_glyphs.c |  24 +++-
 glamor/glamor_copy.c             |  42 ++++++-
 glamor/glamor_dash.c             |  97 +++++++++++++---
 glamor/glamor_program.c          | 190 ++++++++++++++++++++++++++++---
 glamor/glamor_program.h          |   1 +
 glamor/glamor_rects.c            |  22 +++-
 glamor/glamor_spans.c            |  17 ++-
 glamor/glamor_text.c             | 107 ++++++++++++++++-
 9 files changed, 455 insertions(+), 48 deletions(-)

diff --git a/glamor/glamor.c b/glamor/glamor.c
index 7a16afb51..c18fd908a 100644
--- a/glamor/glamor.c
+++ b/glamor/glamor.c
@@ -725,7 +725,8 @@ glamor_init(ScreenPtr screen, unsigned int flags)
          * compatibility features (to reduce the diff between 1.20 and
          * 1.30 programs).
          */
-        glamor_priv->glsl_version = 120;
+        if (gl_version < 30)
+            glamor_priv->glsl_version = 120;
     }
 
     /* We'd like to require GL_ARB_map_buffer_range or
diff --git a/glamor/glamor_composite_glyphs.c b/glamor/glamor_composite_glyphs.c
index c69b940d4..9ab0244d7 100644
--- a/glamor/glamor_composite_glyphs.c
+++ b/glamor/glamor_composite_glyphs.c
@@ -177,6 +177,23 @@ glamor_glyph_add(struct glamor_glyph_atlas *atlas, DrawablePtr glyph_draw)
     return TRUE;
 }
 
+static const glamor_facet glamor_facet_composite_glyphs_300es = {
+    .name = "composite_glyphs",
+    .version = 300,
+    .is_gles = TRUE,
+    .vs_vars = ("in vec4 primitive;\n"
+                "in vec2 source;\n"
+                "out vec2 glyph_pos;\n"),
+    .vs_exec = ("       vec2 pos = primitive.zw * vec2(gl_VertexID&1, (gl_VertexID&2)>>1);\n"
+                GLAMOR_POS(gl_Position, (primitive.xy + pos))
+                "       glyph_pos = (source + pos) * ATLAS_DIM_INV;\n"),
+    .fs_vars = ("in vec2 glyph_pos;\n"
+                "out vec4 myFragColor;\n"),
+    .fs_exec = ("       vec4 mask = texture(atlas, glyph_pos);\n"),
+    .source_name = "source",
+    .locations = glamor_program_location_atlas,
+};
+
 static const glamor_facet glamor_facet_composite_glyphs_130 = {
     .name = "composite_glyphs",
     .version = 130,
@@ -450,7 +467,12 @@ glamor_composite_glyphs(CARD8 op,
                     /* First glyph in the current atlas?
                      */
                     if (_X_UNLIKELY(glyphs_queued == 0)) {
-                        if (glamor_glsl_has_ints(glamor_priv))
+                        if (glamor_priv->is_gles && glamor_priv->glsl_version >= 300)
+                            prog = glamor_setup_program_render(op, src, glyph_pict, dst,
+                                                               glyphs_program,
+                                                               &glamor_facet_composite_glyphs_300es,
+                                                               glamor_priv->glyph_defines);
+                        else if (glamor_glsl_has_ints(glamor_priv))
                             prog = glamor_setup_program_render(op, src, glyph_pict, dst,
                                                                glyphs_program,
                                                                &glamor_facet_composite_glyphs_130,
diff --git a/glamor/glamor_copy.c b/glamor/glamor_copy.c
index 1ab2be6c0..050d2bf8b 100644
--- a/glamor/glamor_copy.c
+++ b/glamor/glamor_copy.c
@@ -57,6 +57,19 @@ static const glamor_facet glamor_facet_copyarea = {
     .use = use_copyarea,
 };
 
+static const glamor_facet glamor_facet_copyarea_300es = {
+    "copy_area",
+    .version = 300,
+    .is_gles = TRUE,
+    .vs_vars = "in vec2 primitive;\n",
+    .vs_exec = (GLAMOR_POS(gl_Position, primitive.xy)
+                "       fill_pos = (fill_offset + primitive.xy) * fill_size_inv;\n"),
+    .fs_vars = "out vec4 myFragColor;\n",
+    .fs_exec = "       myFragColor = texture(sampler, fill_pos);\n",
+    .locations = glamor_program_location_fillsamp | glamor_program_location_fillpos,
+    .use = use_copyarea,
+};
+
 /*
  * Configure the copy plane program for the current operation
  */
@@ -138,7 +151,7 @@ use_copyplane(PixmapPtr dst, GCPtr gc, glamor_program *prog, void *arg)
     return TRUE;
 }
 
-static const glamor_facet glamor_facet_copyplane = {
+static const glamor_facet glamor_facet_copyplane_130 = {
     "copy_plane",
     .version = 130,
     .vs_vars = "attribute vec2 primitive;\n",
@@ -153,6 +166,23 @@ static const glamor_facet glamor_facet_copyplane = {
     .use = use_copyplane,
 };
 
+static const glamor_facet glamor_facet_copyplane_300es = {
+    "copy_plane",
+    .version = 300,
+    .is_gles = TRUE,
+    .vs_vars = "in vec2 primitive;\n",
+    .vs_exec = (GLAMOR_POS(gl_Position, (primitive.xy))
+                "       fill_pos = (fill_offset + primitive.xy) * fill_size_inv;\n"),
+    .fs_vars = "out vec4 myFragColor;\n",
+    .fs_exec = ("       uvec4 bits = uvec4(round(texture(sampler, fill_pos) * bitmul));\n"
+                "       if ((bits & bitplane) != uvec4(0,0,0,0))\n"
+                "               myFragColor = fg;\n"
+                "       else\n"
+                "               myFragColor = bg;\n"),
+    .locations = glamor_program_location_fillsamp|glamor_program_location_fillpos|glamor_program_location_fg|glamor_program_location_bg|glamor_program_location_bitplane,
+    .use = use_copyplane,
+};
+
 /*
  * When all else fails, pull the bits out of the GPU and do the
  * operation with fb
@@ -386,10 +416,16 @@ glamor_copy_fbo_fbo_draw(DrawablePtr src,
 
     if (bitplane) {
         prog = &glamor_priv->copy_plane_prog;
-        copy_facet = &glamor_facet_copyplane;
+        if (glamor_priv->is_gles && glamor_priv->glsl_version >= 300)
+            copy_facet = &glamor_facet_copyplane_300es;
+        else
+            copy_facet = &glamor_facet_copyplane_130;
     } else {
         prog = &glamor_priv->copy_area_prog;
-        copy_facet = &glamor_facet_copyarea;
+        if (glamor_priv->is_gles && glamor_priv->glsl_version >= 300)
+            copy_facet = &glamor_facet_copyarea_300es;
+        else
+            copy_facet = &glamor_facet_copyarea;
     }
 
     if (prog->failed)
diff --git a/glamor/glamor_dash.c b/glamor/glamor_dash.c
index b53ce5c50..2e8f31fa3 100644
--- a/glamor/glamor_dash.c
+++ b/glamor/glamor_dash.c
@@ -26,7 +26,7 @@
 #include "glamor_transfer.h"
 #include "glamor_prepare.h"
 
-static const char dash_vs_vars[] =
+static const char dash_vs_vars_130[] =
     "attribute vec3 primitive;\n"
     "varying float dash_offset;\n";
 
@@ -35,16 +35,16 @@ static const char dash_vs_exec[] =
     "       vec2 pos = vec2(0,0);\n"
     GLAMOR_POS(gl_Position, primitive.xy);
 
-static const char dash_fs_vars[] =
+static const char dash_fs_vars_130[] =
     "varying float dash_offset;\n";
 
-static const char on_off_fs_exec[] =
+static const char on_off_fs_exec_130[] =
     "       float pattern = texture2D(dash, vec2(dash_offset, 0.5)).w;\n"
     "       if (pattern == 0.0)\n"
     "               discard;\n";
 
 /* XXX deal with stippled double dashed lines once we have stippling support */
-static const char double_fs_exec[] =
+static const char double_fs_exec_130[] =
     "       float pattern = texture2D(dash, vec2(dash_offset, 0.5)).w;\n"
     "       if (pattern == 0.0)\n"
     "               gl_FragColor = bg;\n"
@@ -52,23 +52,72 @@ static const char double_fs_exec[] =
     "               gl_FragColor = fg;\n";
 
 
-static const glamor_facet glamor_facet_on_off_dash_lines = {
+static const glamor_facet glamor_facet_on_off_dash_lines_130 = {
     .version = 130,
     .name = "poly_lines_on_off_dash",
-    .vs_vars = dash_vs_vars,
+    .vs_vars = dash_vs_vars_130,
     .vs_exec = dash_vs_exec,
-    .fs_vars = dash_fs_vars,
-    .fs_exec = on_off_fs_exec,
+    .fs_vars = dash_fs_vars_130,
+    .fs_exec = on_off_fs_exec_130,
     .locations = glamor_program_location_dash,
 };
 
-static const glamor_facet glamor_facet_double_dash_lines = {
+static const glamor_facet glamor_facet_double_dash_lines_130 = {
     .version = 130,
     .name = "poly_lines_double_dash",
-    .vs_vars = dash_vs_vars,
+    .vs_vars = dash_vs_vars_130,
     .vs_exec = dash_vs_exec,
-    .fs_vars = dash_fs_vars,
-    .fs_exec = double_fs_exec,
+    .fs_vars = dash_fs_vars_130,
+    .fs_exec = double_fs_exec_130,
+    .locations = (glamor_program_location_dash|
+                  glamor_program_location_fg|
+                  glamor_program_location_bg),
+};
+
+static const char dash_vs_vars_300es[] =
+    "in vec3 primitive;\n"
+    "out float dash_offset;\n";
+
+static const char on_off_fs_vars_300es[] =
+    "in float dash_offset;\n";
+
+static const char on_off_fs_exec_300es[] =
+    "       float pattern = texture(dash, vec2(dash_offset, 0.5)).w;\n"
+    "       if (pattern == 0.0)\n"
+    "               discard;\n";
+
+static const char double_fs_vars_300es[] =
+    "in float dash_offset;\n"
+    "out vec4 myFragColor;\n";
+
+/* XXX deal with stippled double dashed lines once we have stippling support */
+static const char double_fs_exec_300es[] =
+    "       float pattern = texture(dash, vec2(dash_offset, 0.5)).w;\n"
+    "       if (pattern == 0.0)\n"
+    "               myFragColor = bg;\n"
+    "       else\n"
+    "               myFragColor = fg;\n";
+
+
+static const glamor_facet glamor_facet_on_off_dash_lines_300es = {
+    .version = 300,
+    .is_gles = TRUE,
+    .name = "poly_lines_on_off_dash",
+    .vs_vars = dash_vs_vars_300es,
+    .vs_exec = dash_vs_exec,
+    .fs_vars = on_off_fs_vars_300es,
+    .fs_exec = on_off_fs_exec_300es,
+    .locations = glamor_program_location_dash,
+};
+
+static const glamor_facet glamor_facet_double_dash_lines_300es = {
+    .version = 300,
+    .is_gles = TRUE,
+    .name = "poly_lines_double_dash",
+    .vs_vars = dash_vs_vars_300es,
+    .vs_exec = dash_vs_exec,
+    .fs_vars = double_fs_vars_300es,
+    .fs_exec = double_fs_exec_300es,
     .locations = (glamor_program_location_dash|
                   glamor_program_location_fg|
                   glamor_program_location_bg),
@@ -156,9 +205,14 @@ glamor_dash_setup(DrawablePtr drawable, GCPtr gc)
 
     switch (gc->lineStyle) {
     case LineOnOffDash:
-        prog = glamor_use_program_fill(pixmap, gc,
-                                       &glamor_priv->on_off_dash_line_progs,
-                                       &glamor_facet_on_off_dash_lines);
+	if (glamor_priv->is_gles && glamor_priv->glsl_version >= 300)
+            prog = glamor_use_program_fill(pixmap, gc,
+                                           &glamor_priv->on_off_dash_line_progs,
+                                           &glamor_facet_on_off_dash_lines_300es);
+        else
+            prog = glamor_use_program_fill(pixmap, gc,
+                                           &glamor_priv->on_off_dash_line_progs,
+                                           &glamor_facet_on_off_dash_lines_130);
         if (!prog)
             goto bail;
         break;
@@ -169,10 +223,17 @@ glamor_dash_setup(DrawablePtr drawable, GCPtr gc)
         prog = &glamor_priv->double_dash_line_prog;
 
         if (!prog->prog) {
-            if (!glamor_build_program(screen, prog,
-                                      &glamor_facet_double_dash_lines,
-                                      NULL, NULL, NULL))
+	    if (glamor_priv->is_gles && glamor_priv->glsl_version >= 300) {
+                if (!glamor_build_program(screen, prog,
+                                          &glamor_facet_double_dash_lines_300es,
+                                          NULL, NULL, NULL))
+                goto bail;
+            } else {
+                if (!glamor_build_program(screen, prog,
+                                          &glamor_facet_double_dash_lines_130,
+                                          NULL, NULL, NULL))
                 goto bail;
+            }
         }
 
         if (!glamor_use_program(pixmap, gc, prog, NULL))
diff --git a/glamor/glamor_program.c b/glamor/glamor_program.c
index 46a506aaf..08b67241c 100644
--- a/glamor/glamor_program.c
+++ b/glamor/glamor_program.c
@@ -98,6 +98,63 @@ static const glamor_facet *glamor_facet_fill[4] = {
     &glamor_fill_opaque_stipple,
 };
 
+const glamor_facet glamor_fill_solid_300es = {
+    .name = "solid",
+    .version = 300,
+    .is_gles = TRUE,
+    .fs_vars = "out vec4 myFragColor;\n",
+    .fs_exec = "       myFragColor = fg;\n",
+    .locations = glamor_program_location_fg,
+    .use = use_solid,
+};
+
+static const glamor_facet glamor_fill_tile_300es = {
+    .name = "tile",
+    .version = 300,
+    .is_gles = TRUE,
+    .vs_exec =  "       fill_pos = (fill_offset + primitive.xy + pos) * fill_size_inv;\n",
+    .fs_vars = "out vec4 myFragColor;\n",
+    .fs_exec =  "       myFragColor = texture(sampler, fill_pos);\n",
+    .locations = glamor_program_location_fillsamp | glamor_program_location_fillpos,
+    .use = use_tile,
+};
+
+static const glamor_facet glamor_fill_stipple_300es = {
+    .name = "stipple",
+    .version = 300,
+    .is_gles = TRUE,
+    .vs_exec =  "       fill_pos = (fill_offset + primitive.xy + pos) * fill_size_inv;\n",
+    .fs_vars = "out vec4 myFragColor;\n",
+    .fs_exec = ("       float a = texture(sampler, fill_pos).w;\n"
+                "       if (a == 0.0)\n"
+                "               discard;\n"
+                "       myFragColor = fg;\n"),
+    .locations = glamor_program_location_fg | glamor_program_location_fillsamp | glamor_program_location_fillpos,
+    .use = use_stipple,
+};
+
+static const glamor_facet glamor_fill_opaque_stipple_300es = {
+    .name = "opaque_stipple",
+    .version = 300,
+    .is_gles = TRUE,
+    .vs_exec =  "       fill_pos = (fill_offset + primitive.xy + pos) * fill_size_inv;\n",
+    .fs_vars = "out vec4 myFragColor;\n",
+    .fs_exec = ("       float a = texture(sampler, fill_pos).w;\n"
+                "       if (a == 0.0)\n"
+                "               myFragColor = bg;\n"
+                "       else\n"
+                "               myFragColor = fg;\n"),
+    .locations = glamor_program_location_fg | glamor_program_location_bg | glamor_program_location_fillsamp | glamor_program_location_fillpos,
+    .use = use_opaque_stipple
+};
+
+static const glamor_facet *glamor_facet_fill_300es[4] = {
+    &glamor_fill_solid_300es,
+    &glamor_fill_tile_300es,
+    &glamor_fill_stipple_300es,
+    &glamor_fill_opaque_stipple_300es,
+};
+
 typedef struct {
     glamor_program_location     location;
     const char                  *vs_vars;
@@ -144,6 +201,46 @@ static glamor_location_var location_vars[] = {
     },
 };
 
+static glamor_location_var location_vars_300es[] = {
+    {
+        .location = glamor_program_location_fg,
+        .fs_vars = "uniform vec4 fg;\n"
+    },
+    {
+        .location = glamor_program_location_bg,
+        .fs_vars = "uniform vec4 bg;\n"
+    },
+    {
+        .location = glamor_program_location_fillsamp,
+        .fs_vars = "uniform sampler2D sampler;\n"
+    },
+    {
+        .location = glamor_program_location_fillpos,
+        .vs_vars = ("uniform vec2 fill_offset;\n"
+                    "uniform vec2 fill_size_inv;\n"
+                    "out vec2 fill_pos;\n"),
+        .fs_vars = ("in vec2 fill_pos;\n")
+    },
+    {
+        .location = glamor_program_location_font,
+        .fs_vars = "uniform mediump usampler2D font;\n",
+    },
+    {
+        .location = glamor_program_location_bitplane,
+        .fs_vars = ("uniform uvec4 bitplane;\n"
+                    "uniform vec4 bitmul;\n"),
+    },
+    {
+        .location = glamor_program_location_dash,
+        .vs_vars = "uniform float dash_length;\n",
+        .fs_vars = "uniform sampler2D dash;\n",
+    },
+    {
+        .location = glamor_program_location_atlas,
+        .fs_vars = "uniform sampler2D atlas;\n",
+    },
+};
+
 static char *
 add_var(char *cur, const char *add)
 {
@@ -162,26 +259,32 @@ add_var(char *cur, const char *add)
 }
 
 static char *
-vs_location_vars(glamor_program_location locations)
+vs_location_vars(glamor_program_location locations, int version, Bool is_gles)
 {
     int l;
     char *vars = strdup("");
+    glamor_location_var *lvars = (is_gles && version >= 300) ?
+                                     location_vars_300es :
+                                     location_vars;
 
     for (l = 0; vars && l < ARRAY_SIZE(location_vars); l++)
-        if (locations & location_vars[l].location)
-            vars = add_var(vars, location_vars[l].vs_vars);
+        if (locations & lvars[l].location)
+            vars = add_var(vars, lvars[l].vs_vars);
     return vars;
 }
 
 static char *
-fs_location_vars(glamor_program_location locations)
+fs_location_vars(glamor_program_location locations, int version, Bool is_gles)
 {
     int l;
     char *vars = strdup("");
+    glamor_location_var *lvars = (is_gles && version >= 300) ?
+                                     location_vars_300es :
+                                     location_vars;
 
     for (l = 0; vars && l < ARRAY_SIZE(location_vars); l++)
-        if (locations & location_vars[l].location)
-            vars = add_var(vars, location_vars[l].fs_vars);
+        if (locations & lvars[l].location)
+            vars = add_var(vars, lvars[l].fs_vars);
     return vars;
 }
 
@@ -276,7 +379,8 @@ glamor_build_program(ScreenPtr          screen,
     version = MAX(version, fill->version);
 
     if (version > glamor_priv->glsl_version) {
-        if (version == 130 && !glamor_priv->use_gpu_shader4)
+        if (glamor_priv->is_gles ||
+            (version == 130 && !glamor_priv->use_gpu_shader4))
             goto fail;
         else {
             version = 120;
@@ -284,8 +388,8 @@ glamor_build_program(ScreenPtr          screen,
         }
     }
 
-    vs_vars = vs_location_vars(locations);
-    fs_vars = fs_location_vars(locations);
+    vs_vars = vs_location_vars(locations, version, prim->is_gles);
+    fs_vars = fs_location_vars(locations, version, prim->is_gles);
 
     if (!vs_vars)
         goto fail;
@@ -293,8 +397,13 @@ glamor_build_program(ScreenPtr          screen,
         goto fail;
 
     if (version) {
-        if (asprintf(&version_string, "#version %d\n", version) < 0)
-            version_string = NULL;
+        if (prim->is_gles && version >= 300) {
+                if (asprintf(&version_string, "#version %d es\n", version) < 0)
+                    version_string = NULL;
+        } else {
+            if (asprintf(&version_string, "#version %d\n", version) < 0)
+                version_string = NULL;
+        }
         if (!version_string)
             goto fail;
     }
@@ -428,7 +537,10 @@ glamor_use_program_fill(PixmapPtr               pixmap,
         return FALSE;
 
     if (!prog->prog) {
-        fill = glamor_facet_fill[fill_style];
+        if (prim->is_gles && prim->version >= 300)
+            fill = glamor_facet_fill_300es[fill_style];
+        else
+            fill = glamor_facet_fill[fill_style];
         if (!fill)
             return NULL;
 
@@ -585,6 +697,40 @@ static const glamor_facet *glamor_facet_source[glamor_program_source_count] = {
     [glamor_program_source_1x1_picture] = &glamor_source_1x1_picture,
 };
 
+static const glamor_facet glamor_source_solid_300es = {
+    .name = "render_solid",
+    .version = 300,
+    .is_gles = TRUE,
+    .fs_exec = "       vec4 source = fg;\n",
+    .locations = glamor_program_location_fg,
+    .use_render = use_source_solid,
+};
+
+static const glamor_facet glamor_source_picture_300es = {
+    .name = "render_picture",
+    .version = 300,
+    .is_gles = TRUE,
+    .vs_exec =  "       fill_pos = (fill_offset + primitive.xy + pos) * fill_size_inv;\n",
+    .fs_exec =  "       vec4 source = texture(sampler, fill_pos);\n",
+    .locations = glamor_program_location_fillsamp | glamor_program_location_fillpos,
+    .use_render = use_source_picture,
+};
+
+static const glamor_facet glamor_source_1x1_picture_300es = {
+    .name = "render_picture",
+    .version = 300,
+    .is_gles = TRUE,
+    .fs_exec =  "       vec4 source = texture(sampler, vec2(0.5));\n",
+    .locations = glamor_program_location_fillsamp,
+    .use_render = use_source_1x1_picture,
+};
+
+static const glamor_facet *glamor_facet_source_300es[glamor_program_source_count] = {
+    [glamor_program_source_solid] = &glamor_source_solid_300es,
+    [glamor_program_source_picture] = &glamor_source_picture_300es,
+    [glamor_program_source_1x1_picture] = &glamor_source_1x1_picture_300es,
+};
+
 static const char *glamor_combine[] = {
     [glamor_program_alpha_normal]    = "        gl_FragColor = source * mask.a;\n",
     [glamor_program_alpha_ca_first]  = "        gl_FragColor = source.a * mask;\n",
@@ -595,6 +741,12 @@ static const char *glamor_combine[] = {
                                               " gl_SecondaryFragColorEXT = source.a * mask;\n"
 };
 
+static const char *glamor_combine_300es[] = {
+    [glamor_program_alpha_normal]    = "       myFragColor = source * mask.a;\n",
+    [glamor_program_alpha_ca_first]  = "       myFragColor = source.a * mask;\n",
+    [glamor_program_alpha_ca_second] = "       myFragColor = source * mask;\n",
+};
+
 static Bool
 glamor_setup_one_program_render(ScreenPtr               screen,
                                 glamor_program          *prog,
@@ -607,13 +759,21 @@ glamor_setup_one_program_render(ScreenPtr               screen,
         return FALSE;
 
     if (!prog->prog) {
-        const glamor_facet      *fill = glamor_facet_source[source_type];
+        const glamor_facet *fill;
+        const char *combine;
 
-        if (!fill)
+        if (prim->is_gles && prim->version >= 300) {
+            fill = glamor_facet_source_300es[source_type];
+            combine = glamor_combine_300es[alpha];
+        } else {
+            fill = glamor_facet_source[source_type];
+            combine = glamor_combine[alpha];
+        }
+        if (!fill || !combine)
             return FALSE;
 
         prog->alpha = alpha;
-        if (!glamor_build_program(screen, prog, prim, fill, glamor_combine[alpha], defines))
+        if (!glamor_build_program(screen, prog, prim, fill, combine, defines))
             return FALSE;
     }
 
diff --git a/glamor/glamor_program.h b/glamor/glamor_program.h
index 0bd918fff..174672a17 100644
--- a/glamor/glamor_program.h
+++ b/glamor/glamor_program.h
@@ -57,6 +57,7 @@ typedef Bool (*glamor_use_render) (CARD8 op, PicturePtr src, PicturePtr dst, gla
 typedef struct {
     const char                          *name;
     const int                           version;
+    Bool                                is_gles;
     char                                *vs_extensions;
     const char                          *fs_extensions;
     const char                          *vs_vars;
diff --git a/glamor/glamor_rects.c b/glamor/glamor_rects.c
index 8cdad64e4..82668b4d7 100644
--- a/glamor/glamor_rects.c
+++ b/glamor/glamor_rects.c
@@ -24,6 +24,17 @@
 #include "glamor_program.h"
 #include "glamor_transform.h"
 
+static const glamor_facet glamor_facet_polyfillrect_300es = {
+    .name = "poly_fill_rect",
+    .version = 300,
+    .is_gles = TRUE,
+    .source_name = "size",
+    .vs_vars = "in vec2 primitive;\n"
+               "in vec2 size;\n",
+    .vs_exec = ("       vec2 pos = size * vec2(gl_VertexID&1, (gl_VertexID&2)>>1);\n"
+                GLAMOR_POS(gl_Position, (primitive.xy + pos))),
+};
+
 static const glamor_facet glamor_facet_polyfillrect_130 = {
     .name = "poly_fill_rect",
     .version = 130,
@@ -70,9 +81,14 @@ glamor_poly_fill_rect_gl(DrawablePtr drawable,
     }
 
     if (glamor_glsl_has_ints(glamor_priv)) {
-        prog = glamor_use_program_fill(pixmap, gc,
-                                       &glamor_priv->poly_fill_rect_program,
-                                       &glamor_facet_polyfillrect_130);
+        if (glamor_priv->is_gles && glamor_priv->glsl_version >= 300)
+            prog = glamor_use_program_fill(pixmap, gc,
+                                           &glamor_priv->poly_fill_rect_program,
+                                           &glamor_facet_polyfillrect_300es);
+        else
+            prog = glamor_use_program_fill(pixmap, gc,
+                                           &glamor_priv->poly_fill_rect_program,
+                                           &glamor_facet_polyfillrect_130);
 
         if (!prog)
             goto bail;
diff --git a/glamor/glamor_spans.c b/glamor/glamor_spans.c
index 00a019c7b..2dc1f8072 100644
--- a/glamor/glamor_spans.c
+++ b/glamor/glamor_spans.c
@@ -26,6 +26,15 @@
 
 glamor_program  fill_spans_progs[4];
 
+static const glamor_facet glamor_facet_fillspans_300es = {
+    .name = "fill_spans",
+    .version = 300,
+    .is_gles = TRUE,
+    .vs_vars =  "in vec3 primitive;\n",
+    .vs_exec = ("       vec2 pos = vec2(primitive.z,1) * vec2(gl_VertexID&1, (gl_VertexID&2)>>1);\n"
+                GLAMOR_POS(gl_Position, (primitive.xy + pos))),
+};
+
 static const glamor_facet glamor_facet_fillspans_130 = {
     .name = "fill_spans",
     .version = 130,
@@ -65,8 +74,12 @@ glamor_fill_spans_gl(DrawablePtr drawable,
     glamor_make_current(glamor_priv);
 
     if (glamor_glsl_has_ints(glamor_priv)) {
-        prog = glamor_use_program_fill(pixmap, gc, &glamor_priv->fill_spans_program,
-                                       &glamor_facet_fillspans_130);
+        if (glamor_priv->is_gles && glamor_priv->glsl_version >= 300)
+            prog = glamor_use_program_fill(pixmap, gc, &glamor_priv->fill_spans_program,
+                                           &glamor_facet_fillspans_300es);
+        else
+            prog = glamor_use_program_fill(pixmap, gc, &glamor_priv->fill_spans_program,
+                                           &glamor_facet_fillspans_130);
 
         if (!prog)
             goto bail;
diff --git a/glamor/glamor_text.c b/glamor/glamor_text.c
index cf165cad8..d2b0f46d4 100644
--- a/glamor/glamor_text.c
+++ b/glamor/glamor_text.c
@@ -264,6 +264,55 @@ static const glamor_facet glamor_facet_poly_text = {
     .locations = glamor_program_location_font,
 };
 
+static const char vs_vars_text_300es[] =
+    "in vec4 primitive;\n"
+    "in vec2 source;\n"
+    "out vec2 glyph_pos;\n";
+
+static const char vs_exec_text_300es[] =
+    "       vec2 pos = primitive.zw * vec2(gl_VertexID&1, (gl_VertexID&2)>>1);\n"
+    GLAMOR_POS(gl_Position, (primitive.xy + pos))
+    "       glyph_pos = source + pos;\n";
+
+static const char fs_vars_text_300es[] =
+    "in vec2 glyph_pos;\n";
+
+static const char fs_exec_text_300es[] =
+    "       ivec2 itile_texture = ivec2(glyph_pos);\n"
+    "       uint x = uint(itile_texture.x & 7);\n"
+    "       itile_texture.x >>= 3;\n"
+    "       uint texel = texelFetch(font, itile_texture, 0).x;\n"
+    "       uint bit = (texel >> x) & uint(1);\n"
+    "       if (bit == uint(0))\n"
+    "               discard;\n";
+
+static const char fs_vars_te_300es[] =
+    "in vec2 glyph_pos;\n"
+    "out vec4 myFragColor;\n";
+
+static const char fs_exec_te_300es[] =
+    "       ivec2 itile_texture = ivec2(glyph_pos);\n"
+    "       uint x = uint(itile_texture.x & 7);\n"
+    "       itile_texture.x >>= 3;\n"
+    "       uint texel = texelFetch(font, itile_texture, 0).x;\n"
+    "       uint bit = (texel >> x) & uint(1);\n"
+    "       if (bit == uint(0))\n"
+    "               myFragColor = bg;\n"
+    "       else\n"
+    "               myFragColor = fg;\n";
+
+static const glamor_facet glamor_facet_poly_text_300es = {
+    .name = "poly_text",
+    .version = 300,
+    .is_gles = TRUE,
+    .vs_vars = vs_vars_text_300es,
+    .vs_exec = vs_exec_text_300es,
+    .fs_vars = fs_vars_text_300es,
+    .fs_exec = fs_exec_text_300es,
+    .source_name = "source",
+    .locations = glamor_program_location_font,
+};
+
 static Bool
 glamor_poly_text(DrawablePtr drawable, GCPtr gc,
                  int x, int y, int count, char *chars, Bool sixteen, int *final_pos)
@@ -288,7 +337,10 @@ glamor_poly_text(DrawablePtr drawable, GCPtr gc,
 
     glamor_make_current(glamor_priv);
 
-    prog = glamor_use_program_fill(pixmap, gc, &glamor_priv->poly_text_progs, &glamor_facet_poly_text);
+    if (glamor_priv->is_gles && glamor_priv->glsl_version >= 300)
+        prog = glamor_use_program_fill(pixmap, gc, &glamor_priv->poly_text_progs, &glamor_facet_poly_text_300es);
+    else
+        prog = glamor_use_program_fill(pixmap, gc, &glamor_priv->poly_text_progs, &glamor_facet_poly_text);
 
     if (!prog)
         goto bail;
@@ -345,6 +397,18 @@ static const glamor_facet glamor_facet_image_text = {
     .locations = glamor_program_location_font,
 };
 
+static const glamor_facet glamor_facet_image_text_300es = {
+    .name = "image_text",
+    .version = 300,
+    .is_gles = TRUE,
+    .vs_vars = vs_vars_text_300es,
+    .vs_exec = vs_exec_text_300es,
+    .fs_vars = fs_vars_text_300es,
+    .fs_exec = fs_exec_text_300es,
+    .source_name = "source",
+    .locations = glamor_program_location_font,
+};
+
 static Bool
 use_image_solid(PixmapPtr pixmap, GCPtr gc, glamor_program *prog, void *arg)
 {
@@ -358,6 +422,16 @@ static const glamor_facet glamor_facet_image_fill = {
     .use = use_image_solid,
 };
 
+static const glamor_facet glamor_facet_image_fill_300es = {
+    .name = "solid",
+    .version = 300,
+    .is_gles = TRUE,
+    .fs_vars = "out vec4 myFragColor;\n",
+    .fs_exec = "       myFragColor = fg;\n",
+    .locations = glamor_program_location_fg,
+    .use = use_image_solid,
+};
+
 static Bool
 glamor_te_text_use(PixmapPtr pixmap, GCPtr gc, glamor_program *prog, void *arg)
 {
@@ -367,6 +441,19 @@ glamor_te_text_use(PixmapPtr pixmap, GCPtr gc, glamor_program *prog, void *arg)
     return TRUE;
 }
 
+static const glamor_facet glamor_facet_te_text_300es = {
+    .name = "te_text",
+    .version = 300,
+    .is_gles = TRUE,
+    .vs_vars = vs_vars_text_300es,
+    .vs_exec = vs_exec_text_300es,
+    .fs_vars = fs_vars_te_300es,
+    .fs_exec = fs_exec_te_300es,
+    .locations = glamor_program_location_fg | glamor_program_location_bg | glamor_program_location_font,
+    .source_name = "source",
+    .use = glamor_te_text_use,
+};
+
 static const glamor_facet glamor_facet_te_text = {
     .name = "te_text",
     .version = 130,
@@ -416,11 +503,21 @@ glamor_image_text(DrawablePtr drawable, GCPtr gc,
 
     if (!prog->prog) {
         if (TERMINALFONT(gc->font)) {
-            prim_facet = &glamor_facet_te_text;
-            fill_facet = NULL;
+            if (glamor_priv->is_gles && glamor_priv->glsl_version >= 300) {
+                prim_facet = &glamor_facet_te_text_300es;
+                fill_facet = NULL;
+            } else {
+                prim_facet = &glamor_facet_te_text;
+                fill_facet = NULL;
+            }
         } else {
-            prim_facet = &glamor_facet_image_text;
-            fill_facet = &glamor_facet_image_fill;
+            if (glamor_priv->is_gles && glamor_priv->glsl_version >= 300) {
+                prim_facet = &glamor_facet_image_text_300es;
+                fill_facet = &glamor_facet_image_fill_300es;
+            } else {
+                prim_facet = &glamor_facet_image_text;
+                fill_facet = &glamor_facet_image_fill;
+            }
         }
 
         if (!glamor_build_program(screen, prog, prim_facet, fill_facet, NULL, NULL))
-- 
2.34.1

